source/module_hsolver/diago_cg.cpp

```diff
@@ -683,23 +683,37 @@ void DiagoCG<FPTYPE, Device>::diag(hamilt::Hamilt<FPTYPE, Device> *phm_in, psi::
     {
         if(DiagoIterAssist<FPTYPE>::need_subspace || ntry > 0)
         {
-            #if defined(__CUDA) || defined(__ROCM)
-            psi::Psi<std::complex<FPTYPE>, psi::DEVICE_CPU> cpu_psi = psi;
-            // hamilt::Hamilt<FPTYPE>* h_phm_in = new hamilt::HamiltPW<FPTYPE>(reinterpret_cast<hamilt::HamiltPW<FPTYPE>*>(phm_in));
-            hamilt::Hamilt<FPTYPE, psi::DEVICE_CPU>* h_phm_in =
-                    new hamilt::HamiltPW<FPTYPE, psi::DEVICE_CPU>(
-                            reinterpret_cast<hamilt::HamiltPW<FPTYPE, psi::DEVICE_GPU>*>(phm_in));
-            DiagoIterAssist<FPTYPE>::diagH_subspace(h_phm_in, cpu_psi, cpu_psi, eigenvalue_in);
-            psi::memory::synchronize_memory_op<std::complex<FPTYPE>, Device, psi::DEVICE_CPU>()(
-                    psi.get_device(),
-                    cpu_psi.get_device(),
-                    psi.get_pointer() - psi.get_psi_bias(),
-                    cpu_psi.get_pointer() - cpu_psi.get_psi_bias(),
-                    psi.size());
-            delete reinterpret_cast<hamilt::HamiltPW<FPTYPE, psi::DEVICE_CPU>*>(h_phm_in);
-            #else
-            DiagoIterAssist<FPTYPE>::diagH_subspace(phm_in, psi, psi, eigenvalue_in);
-            #endif
+        #if defined(__CUDA) || defined(__ROCM)
```

Comment:
Please consider move those eigenvalue_gpu setting into the diagH_subspace dunction to keep the interface consistant with the CPU in diago_cg. We are not using the GPU memory to store the eigen values anywhere.

Reply:
Or maybe we could involve a params to control the location of output eigenvalues

Reply:
Because the cusolver interface called in the GPU version of the diagH_LAPACK requires that the array of eigenvalues be stored in the GPU. I can simply move this memory transfer from outside the diagH_subspace function to inside the diagH_LAPACK function, but I can't avoid the memory transfer here.